### PR TITLE
Adding dev-spaces extension under the aks module

### PR DESCRIPTION
**Pull request by saurabsa**
use-dev-spaces
 - Installs AZDS cli tool (if doesnot exist)
 - Creates and Selects Dev Spaces resource
remove-dev-spaces
 - Deletes the resource

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `./scripts/ci/test_static.sh` locally? (`pip install pylint flake8` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli-extensions/blob/master/docs/extension_summary_guidelines.md).

### CODE REVIEW

File: src/dev-spaces/setup.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    'Programming Language :: Python :: 3.6',
+    'License :: OSI Approved :: MIT License',
+]
+
+DEPENDENCIES = []
+
+setup(
+    name='dev-spaces',
+    version=VERSION,
+    description='An Azure CLI Extension to manage dev spaces',
```
Comment (derekbekoe) [inline src/dev-spaces/setup.py]: Change the description to not say 'An Azure CLI Extension'.
See https://github.com/Azure/azure-cli-extensions/blob/master/docs/extension_summary_guidelines.md

File: src/dev-spaces-preview/azext_dev_spaces_preview/__init__.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            c.argument('cluster_name', options_list=['--name', '-n'])
+            c.argument('resource_group_name', options_list=['--resource-group', '-g'])
+            c.argument('space_name', options_list=['--space', '-s'])
+            c.argument('parent_space_name', options_list=['--parent-space'])
+
+        with self.argument_context('aks remove-dev-spaces') as c:
+            c.argument('cluster_name', options_list=['--name', '-n'])
+            c.argument('resource_group_name', options_list=['--resource-group', '-g'])
+            c.argument('prompt', options_list=['--yes', '-y'], action='store_true',
+                       help='If true, it does not prompt for confirmation.')
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/__init__.py]: The standard help text for this option is "Do not prompt for confirmation."

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+helps['aks use-dev-spaces'] = """
+    type: command
+    short-summary: (PREVIEW) Use Azure Dev Spaces with a managed Kubernetes cluster.
+    parameters:
+        - name: --name -n
+          type: string
+          short-summary: Name of the managed cluster.
+        - name: --resource-group -g
+          type: string
+          short-summary: Name of resource group. You can configure the default group. Using `az configure –defaults group=<name>`.
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: I'm not sure the CLI automatically wraps and aligns help text - can you check? There is supposed to be a line break between "group" and "using". Also there shouldn't be a period after "default group." Finally "Using" shouldn't be capitalized.

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        - name: --space -s
+          type: string
+          short-summary: Name of the dev space to use.
+        - name: --parent-space
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: Let's add -p as a shortcut for this option.

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+          type: string
+          short-summary: Name of a parent dev space to inherit from when creating a new dev space. By default, if there is already a single dev space with no parent, the new space inherits from this one.
+"""
+
+helps['aks remove-dev-spaces'] = """
+    type: command
+    short-summary: (PREVIEW) Detach Azure Dev Spaces from a managed Kubernetes cluster.
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: Sorry, I changed the help here to **Remove** Azure Dev Spaces to match the verb of the command.

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    parameters:
+        - name: --name -n
+          type: string
+          short-summary: Name of the managed cluster.
+        - name: --resource-group -g
+          type: string
+          short-summary: Name of resource group. You can configure the default group. Using `az configure –defaults group=<name>`.
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: Same comment on this as above.

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        - name: --yes -y
+          type: bool
+          short-summary: Do not prompt for confirmation.
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: Ah, so this is correct. Where does the help text in the other file show up then?

File: src/dev-spaces-preview/azext_dev_spaces_preview/custom.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    :type parent_space_name: String
+    """
+
+    azds_tool = 'Azure Dev Spaces CLI (Preview)'
+    should_install_vsce = False
+    system = platform.system()
+    if system == 'Windows':
+        # Windows
+        # Dev Connect Install Path (WinX)
+        azds_cli = os.path.join(os.environ["ProgramW6432"],
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: In order to install, does the user need to be admin? What if they run az not as admin? Are they forced to elevate? That's a little weird.

Comment (stepro) [reply src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: We really should be installing a private copy of the CLI inside where our az extension is installed, which is under the user's ~/.azure/cliextensions/dev-spaces-preview directory.

File: src/dev-spaces-preview/azext_dev_spaces_preview/custom.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    elif system == 'Darwin':
+        # OSX
+        azds_cli = 'azds'
+        setup_file = os.path.join(_create_tmp_dir(), 'azds-osx-setup.sh')
+        setup_url = "https://aka.ms/get-azds-osx-az"
+        setup_args = ['bash', setup_file]
+    else:
+        raise CLIError('Platform not supported: {}.'.format(system))
+
+    should_install_vsce = not _is_dev_connect_installed(azds_cli)
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: _is_dev_spaces_installed

File: src/dev-spaces-preview/azext_dev_spaces_preview/custom.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+    if should_install_vsce:
+        # Install VSCE
+        logger.info('Installing Dev Spaces (Preview) commands...')
+        from six.moves.urllib.request import urlretrieve
+        urlretrieve(setup_url, setup_file)
+        try:
+            subprocess.call(
+                setup_args, universal_newlines=True, stdin=None, stdout=None, stderr=None, shell=False)
+        except PermissionError as ex:
```
Comment (stepro) [inline src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: Yeah, we really can't do this...

File: src/dev-spaces-preview/azext_dev_spaces_preview/custom.py (hunk 8 above)
Comment (stepro) [reply src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: Perhaps we have to defer this until all the remaining commands are in the CLI.

File: src/dev-spaces-preview/setup.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    name='dev-spaces-preview',
+    version=VERSION,
+    description='Dev Spaces provides a rapid, iterative Kubernetes development experience for teams.',
+    long_description='Iteratively develop and debug containers in a Azure Kubernetes Service cluster using Dev Spaces. \
+    Share an AKS cluster with your team and collaborate together. You can test code end-to-end without replicating or \
+    mocking up dependencies. Onboard new team members faster by minimizing their local dev machine setup and have them \
+    work in a consistent dev environment.',
+    license='MIT',
+    author='Microsoft Corporation',
+    author_email='mindaro@microsoft.com',
```
Comment (stepro) [inline src/dev-spaces-preview/setup.py]: Shouldn't be exposing this name...

File: src/dev-spaces-preview/azext_dev_spaces_preview/__init__.py (hunk 2 above)
Comment (saurabsa) [reply src/dev-spaces-preview/azext_dev_spaces_preview/__init__.py]: This is a long summary. The description is already there in _help.py

File: src/dev-spaces-preview/readme.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# Azure CLI Storage Preview Extension #
```
Comment (derekbekoe) [inline src/dev-spaces-preview/readme.md]: The title is wrong

File: src/dev-spaces-preview/setup.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+    'Programming Language :: Python :: 3.4',
+    'Programming Language :: Python :: 3.5',
+    'Programming Language :: Python :: 3.6',
+    'License :: OSI Approved :: MIT License',
+]
+
+DEPENDENCIES = []
+
+setup(
+    name='dev-spaces-preview',
```
Comment (derekbekoe) [inline src/dev-spaces-preview/setup.py]: It's debatable whether you need `-preview` in the name when it's already marked as preview with the metadata.
I'll leave this up to you though.

File: src/dev-spaces-preview/azext_dev_spaces_preview/_help.py (hunk 7 above)
Comment (saurabsa) [reply src/dev-spaces-preview/azext_dev_spaces_preview/_help.py]: I think that is a redundancy, I will remove it.

File: src/dev-spaces-preview/azext_dev_spaces_preview/custom.py (hunk 8 above)
Comment (saurabsa) [reply src/dev-spaces-preview/azext_dev_spaces_preview/custom.py]: Currently, we are going with what our usual setup scripts go and using the same location. We can defer it for later and update it then.

File: src/dev-spaces-preview/setup.py (hunk 13 above)
Comment (saurabsa) [reply src/dev-spaces-preview/setup.py]: We had a discussion on this and would like to keep it for now.

**Comment by williexu** [comment]
@saurabsa almost there, just add yourself as codeowner :)

**Pull request merged**